src/XamlX.IL.Cecil/CecilTypeBuilder.cs

```diff
@@ -11,11 +11,11 @@ partial class CecilTypeSystem
     {
         class CecilTypeBuilder : CecilType, IXamlTypeBuilder<IXamlILEmitter>
         {
-            protected TypeReference SelfReference;
-            public CecilTypeBuilder(CecilTypeSystem typeSystem, CecilAssembly assembly, TypeDefinition definition) 
-                : base(typeSystem, assembly, definition)
+            private CecilTypeResolver _builderTypeResolver;
+            public CecilTypeBuilder(CecilTypeResolver parentTypeResolver, CecilAssembly assembly, TypeDefinition definition) 
+                : base(parentTypeResolver, assembly, definition)
             {
-                SelfReference = definition;
+                _builderTypeResolver = parentTypeResolver.Nested(definition);
```

Comment:
SelfReference type was replaced with a `_builderTypeResolver`. In this context, it has the same meaning, as all nested types will be resolved from it, like they did previously from SelfReference.